server/data_soma/soma_adaptor.py

```diff
@@ -0,0 +1,339 @@
+        else:
+            df = self.data.var.df()
+            df.insert(0, self.schema["annotations"]["var"]["index"], self.data.var.df().index)
+
+        if fields is not None and len(fields) > 0:
+            df = df[fields]
+
+        df.index = range(df.shape[0])
+        return encode_matrix_fbs(df, col_idx=df.columns)
+
+    def compute_diffexp_ttest(self, maskA, maskB, top_n, lfc_cutoff):
+        if top_n is None:
+            top_n = self.dataset_config.diffexp__top_n
+        if lfc_cutoff is None:
+            lfc_cutoff = self.dataset_config.diffexp__lfc_cutoff
+        return diffexp_generic.diffexp_ttest(self, maskA, maskB, top_n, lfc_cutoff)
+
+    def _is_valid_layout(self, arr):
+        """return True if this layout data is a valid array for front-end presentation:
+        * ndarray, dtype float/int/uint
+        * with shape (n_obs, >= 2)
+        * with all values finite or NaN (no +Inf or -Inf)
+        """
+        is_valid = type(arr) == np.ndarray and arr.dtype.kind in "fiu"
+        is_valid = is_valid and arr.shape[0] == self.data.obs.shape()[0] and arr.shape[1] >= 2
+        is_valid = is_valid and not np.any(np.isinf(arr)) and not np.all(np.isnan(arr))
+        return is_valid
+
+    def _load_data(self, data_locator):
+        try:
+            # there is no guarantee data_locator indicates a local file. If we get a non-local object,
+            # make a copy in tmp, and delete it after we load into memory.
+            with data_locator.local_handle() as lh:
+                self.data = tiledbsc.SOMA(lh)
+
+        except ValueError:
+            raise DatasetAccessError("Data must be in the SOMA format.")
+        except MemoryError:
+            raise DatasetAccessError("Out of memory - file is too large for available memory.")
+        except Exception:
+            raise DatasetAccessError(
+                "Folder not found or is inaccessible. Folder must be in SOMA format."
+                "Please check your input and try again."
+            )
+
+    def is_data_unique(self):
+        obs_names = self.get_obs_keys()
+        var_names = self.get_var_keys()
+        return len(set(obs_names)) == len(obs_names) and len(set(var_names)) == len(var_names)
+
+    def _validate_and_initialize(self):
+        if not self.is_data_unique:
+            raise KeyError("All annotation column names must be unique.")
+
+        shape = self.get_shape()
+        self._alias_annotation_names()
+        self.cell_count = shape[0]
+        self.gene_count = shape[1]
+        self._create_schema()
+
+        if self.dataset_config.X_approximate_distribution == "auto":
+            self.get_X_approximate_distribution()
+        else:
+            self.X_approximate_distribution = self.dataset_config.X_approximate_distribution
+
+        # heuristic
+        n_values = shape[0] * shape[1]
+        if n_values > 5e8:
+            self.parameters.update({"diffexp_may_be_slow": True})
+
+    def _alias_annotation_names(self):
```

Comment:
There is one necessary difference, so I can't reuse the exact same function. I could factor out part of it, but I would have to pass in all the class attributes if it were in a utility class or move it to `data_adaptor.py`. But if such functionality can be part of `data_adaptor.py` design-wise, why was it put in `anndata_adaptor.py` only to start?

Reply:
Similar to above, I lack historical context and will defer to Trent/Emanuele.

Reply:
Factor out what you can into data_adaptor. We had other data formats supported that were depreciated. This was not initially included in data-adaptor because it was specific to anndata.